### PR TITLE
Move ItemCollection and derived classes to core

### LINKED ISSUE #1397: Move Sidre collections to core
Sidre's map and item collections can be used in other components:

- https://github.com/LLNL/axom/blob/develop/src/axom/sidre/core/MapCollection.hpp
- https://github.com/LLNL/axom/blob/develop/src/axom/sidre/core/ItemCollection.hpp

Since those classes are not specific to Axom and have no other Axom dependencies, it makes sense to move them to the Axom core component.

### PR DESCRIPTION
**Pull request by nselliott**
# Summary

- This PR is a refactoring
- It does the following:
  - Moves the ItemCollection and its child classes from Sidre to core, to enable them to be used without depending on the Sidre component.
  - Fixes #1397 , may be useful within Sina

Moving these header-only classes required removing their dependencies on Slic, as there were a few uses of Slic assertions and warnings.  Those have been replaced with raw assert() calls and output to std::cerr, matching what is done elsewhere in Axom core.

The definitions of constants InvalidIndex and InvalidName needed to be moved from Sidre into core for usage by the collection classes.

### CODE REVIEW

**Review by rhornung67 — APPROVED** [review]
Thanks @nselliott

File: src/axom/core/MapCollection.hpp [hunk 1]
```diff
@@ -222,7 +220,7 @@ class MapCollection : public ItemCollection<T>
   const std::string& getItemName(IndexType idx) const
   {
     return (hasItem(idx) ? m_items[static_cast<unsigned>(idx)]->getName()
```
Comment (kennyweiss) [inline src/axom/core/MapCollection.hpp:222]: @nselliott -- This function requires the templated type `T` to have a `getName()` function.  

Could you please document this requirement?
Alternatively, is there a way to conditionally include this function via `enable_if` ?

**Review by kennyweiss — APPROVED** [review]
Thanks @nselliott 

Please update the RELEASE_NOTES with this change.
Should add some unit tests to `core` for basic operations on these classes?

File: src/axom/core/MapCollection.hpp (hunk 1 above)
Comment (nselliott) [reply src/axom/core/MapCollection.hpp:222]: This requirement is documented in the header.  I think it's better not to add a SFINAE implementation, as the core feature of `MapCollection` is to hold a collection of named items.  Using `enable_if` would make the code more complicated and allow instances of `MapCollection` to compile that don't have a meaningful use case, so in my opinion it's better to catch attempts to do that at compile time.

**Comment by nselliott** [comment]
> Thanks @nselliott
> 
> Please update the RELEASE_NOTES with this change. Should add some unit tests to `core` for basic operations on these classes?

The existing set of unit tests for these classes in the sidre test directory is still good.  We could move them to the core test directory if needed.

**Pull request merged**